### PR TITLE
Reapply "Native assets: roll deps" (#141748)

### LINKED ISSUE #141827: rolling `native_assets_cli` fails
When I tried to roll this dependency, it failed to compile the tool with:
```
flutter/packages/flutter_tools/lib/src/native_assets.dart:135:27: Error: The argument type 'LinkModePreference/*1*/' can't be assigned to the parameter type 'LinkModePreference/*2*/'.
 - 'LinkModePreference/*1*/' is from 'package:native_assets_cli/src/api/link_mode_preference.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/api/link_mode_preference.dart').
 - 'LinkModePreference/*2*/' is from 'package:native_assets_cli/src/model/link_mode_preference.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/model/link_mode_preference.dart').
      linkModePreference: linkModePreference,
                          ^
flutter/packages/flutter_tools/lib/src/native_assets.dart:136:17: Error: The argument type 'OS/*1*/' can't be assigned to the parameter type 'OS/*2*/'.
 - 'OS/*1*/' is from 'package:native_assets_cli/src/api/target.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/api/target.dart').
 - 'OS/*2*/' is from 'package:native_assets_cli/src/model/target.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/model/target.dart').
      targetOs: targetOS,
                ^
flutter/packages/flutter_tools/lib/src/native_assets.dart:158:18: Error: The argument type 'BuildMode/*1*/' can't be assigned to the parameter type 'BuildMode/*2*/'.
 - 'BuildMode/*1*/' is from 'package:native_assets_cli/src/api/build_mode.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/api/build_mode.dart').
 - 'BuildMode/*2*/' is from 'package:native_assets_cli/src/model/build_mode.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/model/build_mode.dart').
      buildMode: buildMode,
                 ^
flutter/packages/flutter_tools/lib/src/native_assets.dart:159:24: Error: The argument type 'CCompilerConfig/*1*/?' can't be assigned to the parameter type 'CCompilerConfig/*2*/?'.
 - 'CCompilerConfig/*1*/' is from 'package:native_assets_cli/src/api/build_config.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/api/build_config.dart').
 - 'CCompilerConfig/*2*/' is from 'package:native_assets_cli/src/model/build_config.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/model/build_config.dart').
      cCompilerConfig: cCompilerConfig,
                       ^
flutter/packages/flutter_tools/lib/src/native_assets.dart:161:27: Error: The argument type 'LinkModePreference/*1*/' can't be assigned to the parameter type 'LinkModePreference/*2*/'.
 - 'LinkModePreference/*1*/' is from 'package:native_assets_cli/src/api/link_mode_preference.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/api/link_mode_preference.dart').
 - 'LinkModePreference/*2*/' is from 'package:native_assets_cli/src/model/link_mode_preference.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/model/link_mode_preference.dart').
      linkModePreference: linkModePreference,
                          ^
flutter/packages/flutter_tools/lib/src/native_assets.dart:162:15: Error: The argument type 'Target/*1*/' can't be assigned to the parameter type 'Target/*2*/'.
 - 'Target/*1*/' is from 'package:native_assets_cli/src/api/target.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/api/target.dart').
 - 'Target/*2*/' is from 'package:native_assets_cli/src/model/target.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/model/target.dart').
      target: target,
              ^
flutter/packages/flutter_tools/lib/src/native_assets.dart:164:21: Error: The argument type 'IOSSdk/*1*/?' can't be assigned to the parameter type 'IOSSdk/*2*/?'.
 - 'IOSSdk/*1*/' is from 'package:native_assets_cli/src/api/ios_sdk.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/api/ios_sdk.dart').
 - 'IOSSdk/*2*/' is from 'package:native_assets_cli/src/model/ios_sdk.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/model/ios_sdk.dart').
      targetIOSSdk: targetIOSSdk,
                    ^
flutter/packages/flutter_tools/lib/src/native_assets.dart:200:50: Error: The method 'toNativeAssetsFile' isn't defined for the class 'Iterable<Asset>'.
 - 'Iterable' is from 'dart:core'.
 - 'Asset' is from 'package:native_assets_cli/src/api/asset.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/api/asset.dart').
Try correcting the name to the name of an existing method, or defining a method named 'toNativeAssetsFile'.
  final String nativeAssetsDartContents = assets.toNativeAssetsFile();
                                                 ^^^^^^^^^^^^^^^^^^
flutter/packages/flutter_tools/lib/src/native_assets.dart:295:53: Error: The method 'whereLinkMode' isn't defined for the class 'List<Asset>'.
 - 'List' is from 'dart:core'.
 - 'Asset' is from 'package:native_assets_cli/src/api/asset.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/api/asset.dart').
Try correcting the name to the name of an existing method, or defining a method named 'whereLinkMode'.
  final Iterable<Asset> staticAssets = nativeAssets.whereLinkMode(LinkMode.static);
                                                    ^^^^^^^^^^^^^
flutter/packages/flutter_tools/lib/src/native_assets.dart:638:20: Error: The method 'copyWith' isn't defined for the class 'Asset'.
 - 'Asset' is from 'package:native_assets_cli/src/api/asset.dart' ('/home/chrome-bot/.pub-cache/hosted/pub.dev/native_assets_cli-0.4.1/lib/src/api/asset.dart').
Try correcting the name to the name of an existing method, or defining a method named 'copyWith'.
      return asset.copyWith(path: AssetAbsolutePath(uri));
                   ^^^^^^^^
```

### PR DESCRIPTION
**Pull request by dcharkes**
Fixes https://github.com/flutter/flutter/issues/141827

Reland: https://dart-review.googlesource.com/c/sdk/+/346960 has rolled into g3, so the imports should now resolve in g3 as well.

> [!CAUTION]
> _Do NOT merge if "Google Testing" bot didn't run!_

Rolls the packages from https://github.com/dart-lang/native in the native assets implementation.

Most notable we're refactoring `package:native_assets_cli` for `build.dart` use.
Therefore, all imports to that package for Flutter/Dart should be to the implementation internals that are no longer visible for `build.dart` writers. Hence all the import updates.

No behavior in Flutter apps should change.

This PR also updates the template to use the latests version of `package:native_assets_cli` which no longer exposes all the implementation details.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by dcharkes** [comment]
@christopherfujino The "Google Testing" bot again isn't running?

**Comment by christopherfujino** [comment]
> @christopherfujino The "Google Testing" bot again isn't running?

I chatted with Casey about this, and it appears to be broken. I am filing a bug now...

**Comment by CaseyHillers** [comment]
\cc @Jasguerrero FYI. This is on the PRs blocked from not having testing.

**Comment by dcharkes** [comment]
I don't feel very comfortable hitting the submit without testing.

FYI Landing this PR is blocking: https://dart-review.googlesource.com/c/sdk/+/346761 (The cbuild shows that flutter_tools starts failing to build in g3 if we try to land that. Due to the import paths in flutter_tools not yet having changed.) So, I want to make progress on this. Please keep me updated. And if for some reason we can't test the roll with FROB is there a way for me to test it manually?

**Comment by christopherfujino** [comment]
> Thanks @christopherfujino!
> 
> I don't feel very comfortable hitting the submit without testing.
> 
> FYI Landing this PR is blocking: https://dart-review.googlesource.com/c/sdk/+/346761 (The cbuild shows that flutter_tools starts failing to build in g3 if we try to land that. Due to the import paths in flutter_tools not yet having changed.) So, I want to make progress on this. Please keep me updated. And if for some reason we can't test the roll with FROB is there a way for me to test it manually?

My understanding is that getting this unstuck is as simple as manually running a db command (which I believe Casey ran), and that once the CRON runs again it will detect this PR and schedule Google testing. Do I have that right @CaseyHillers?

**Comment by christopherfujino** [comment]
Actually per b/321251334#comment2, it looks like this is still broken and assigned to @Jasguerrero

**Comment by CaseyHillers** [comment]
Following up here, @dcharkes is creating a local version of the change and testing it now. We expect FRoB should recover in the next few hours.

**Comment by dcharkes** [comment]
image

It looks like FRoB has recovered.

File: packages/flutter_tools/lib/src/update_packages_pins.dart [hunk 1]
```diff
@@ -28,6 +28,6 @@ const Map<String, String> kManuallyPinnedDependencies = <String, String>{
   'test_api': '0.6.1', // https://github.com/flutter/flutter/issues/140169
   'test_core': '0.5.9', // https://github.com/flutter/flutter/issues/140169
   'test': '1.24.9', // https://github.com/flutter/flutter/issues/140169
-  'native_assets_builder': '0.3.0', // https://github.com/flutter/flutter/pull/141814
-  'native_assets_cli': '0.3.2', // https://github.com/flutter/flutter/pull/141814
+  'native_assets_builder': '0.3.2', // https://github.com/flutter/flutter/pull/141814
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/update_packages_pins.dart:31]: I think we can just remove these pins since you're updating the tool code in this PR.

**Review by christopherfujino — APPROVED** [review]
LGTM

**Comment by dcharkes** [comment]
> [Mon, 22 Jan 2024 10:11:11 UTC] The file changes in this PR do not need to be rolled into google codebase

Well, now that is terrifying! @CaseyHillers our detection of what needs to be tested might be off.

I've run the command locally as well, and I don't see any g3 errors related to flutter_tools, so I think it's safe to land.

**Pull request merged**